File: backend/src/seed.ts
import { randomUUID } from "node:crypto";

import { faker } from "@faker-js/faker";
import { Decimal } from "decimal.js";

import { PgFxRepository } from "./database/repos/pgFx.js";
import { FxService } from "./services/fx.js";
import {
  COUNTRY_CODES,
  COUNTRY_CURRENCY,
  DEPARTMENTS,
} from "./common/constants.js";
import { migrate, pool } from "./db.js";

const TOTAL = 10_000;
const SEED = 42;

const COUNTRY_BASE: Record<string, string> = {
  US: "120000",
  GB: "75000",
  DE: "80000",
  NL: "78000",
  SG: "110000",
  AU: "115000",
  CA: "105000",
  JP: "8500000",
  AE: "220000",
  IN: "1800000",
};

const DEPT_FACTOR: Record<string, string> = {
  Engineering: "1.15",
  Product: "1.10",
  Design: "1.00",
  Sales: "1.05",
  Marketing: "0.95",
  Finance: "1.00",
  HR: "0.90",
  Operations: "0.85",
  "Customer Success": "0.88",
  Legal: "1.12",
};

const TITLES: Record<string, string[]> = {
  Engineering: ["Software Engineer", "Senior Engineer", "Staff Engineer", "Engineering Manager"],
  Product: ["Product Manager", "Senior Product Manager", "Product Lead"],
  Design: ["Product Designer", "Senior Designer", "Design Lead"],
  Sales: ["Account Executive", "Senior AE", "Sales Manager"],
  Marketing: ["Marketing Manager", "Content Lead", "Growth Manager"],
  Finance: ["Financial Analyst", "Accountant", "Finance Manager"],
  HR: ["HR Generalist", "HRBP", "People Partner"],
  Operations: ["Operations Associate", "Ops Manager"],
  "Customer Success": ["CSM", "Senior CSM"],
  Legal: ["Counsel", "Senior Counsel"],
};

function pick<T>(items: readonly T[], rng: () => number): T {
  return items[Math.floor(rng() * items.length)] as T;
}

function mulberry32(seed: number) {
  return () => {
    let t = (seed += 0x6d2b79f5);
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

function amount(country: string, department: string, rng: () => number): Decimal {
  const mid = new Decimal(COUNTRY_BASE[country]).mul(DEPT_FACTOR[department]);
  return mid.mul(new Decimal((0.72 + rng() * 0.73).toFixed(4))).toDecimalPlaces(0);
}

function addDays(iso: string, days: number): string {
  const date = new Date(`${iso}T00:00:00Z`);
  date.setUTCDate(date.getUTCDate() + days);
  return date.toISOString().slice(0, 10);
}

async function insert(
  sql: string,
  rows: unknown[][],
) {
  const batch = 200;
  for (let i = 0; i < rows.length; i += batch) {
    const slice = rows.slice(i, i + batch);
    const values: string[] = [];
    const params: unknown[] = [];
    for (const row of slice) {
      const start = params.length;
      values.push(`(${row.map((_, idx) => `$${start + idx + 1}`).join(",")})`);
      params.push(...row);
    }
    await pool.query(`${sql} VALUES ${values.join(",")}`, params);
  }
}

async function run() {
  await migrate();
  const existing = await pool.query<{ count: number }>("SELECT COUNT(*)::int AS count FROM employees");
  if ((existing.rows[0]?.count ?? 0) > 0 && !process.argv.includes("--force")) {
    console.log("Database already has employees. Refusing to seed twice. Pass --force to replace.");
    return;
  }
  if (process.argv.includes("--force")) {
    await pool.query("TRUNCATE salary_records, employees CASCADE");
  }

  await new FxService(new PgFxRepository(pool)).seedDefaults();

  const rng = mulberry32(SEED);
  faker.seed(SEED);
  const asOf = "2026-08-29";
  const countryWeights = [28, 22, 10, 8, 6, 6, 6, 5, 5, 4];
  const weightSum = countryWeights.reduce((a, b) => a + b, 0);

  const employees: unknown[][] = [];
  const salaries: unknown[][] = [];

  for (let index = 1; index <= TOTAL; index += 1) {
    const countryRoll = rng() * weightSum;
    let acc = 0;
    let country: (typeof COUNTRY_CODES)[number] = COUNTRY_CODES[0];
    for (let i = 0; i < COUNTRY_CODES.length; i += 1) {
      acc += countryWeights[i];
      if (countryRoll <= acc) {
        country = COUNTRY_CODES[i];
        break;
      }
    }
    const department = pick(DEPARTMENTS, rng);
    const now = new Date();
    const id = randomUUID();
    employees.push([
      id,
      `ACME-${String(index).padStart(5, "0")}`,
      faker.person.fullName(),
      `${faker.internet.username()}.${index}@acme.test`.toLowerCase(),
      country,
      department,
      pick(TITLES[department], rng),
      rng() < 0.08 ? "inactive" : "active",
      now,
      now,
    ]);

    if (rng() < 0.02) {
      continue;
    }

    const hired = addDays(asOf, -Math.floor(30 + rng() * 365 * 8));
    const currency = COUNTRY_CURRENCY[country];
    let lastFrom = addDays(hired, Math.floor(rng() * 60));
    let lastAmount = amount(country, department, rng);
    salaries.push([
      randomUUID(),
      id,
      lastAmount.toFixed(2),
      currency,
      lastFrom,
      "offer accepted",
      now,
    ]);

    const extra = rng() < 0.55 ? 0 : rng() < 0.78 ? 1 : 2;
    for (let step = 0; step < extra; step += 1) {
      lastFrom = addDays(lastFrom, Math.floor(280 + rng() * 140));
      if (lastFrom > addDays(asOf, 180)) {
        break;
      }
      lastAmount = lastAmount.mul(new Decimal((1.04 + rng() * 0.14).toFixed(4))).toDecimalPlaces(0);
      salaries.push([
        randomUUID(),
        id,
        lastAmount.toFixed(2),
        currency,
        lastFrom,
        lastFrom > asOf
          ? "scheduled raise"
          : pick(["annual revision", "promotion", "market adjustment"], rng),
        now,
      ]);
    }
  }

  await insert(
    `INSERT INTO employees
      (id, employee_code, name, email, country_code, department, designation, status, created_at, updated_at)`,
    employees,
  );
  await insert(
    `INSERT INTO salary_records
      (id, employee_id, base_amount, currency, effective_from, reason, created_at)`,
    salaries,
  );

  console.log(`Seeded ${employees.length} employees and ${salaries.length} salary records.`);
  await pool.end();
}

run().catch(async (error) => {
  console.error(error);
  await pool.end();
  process.exit(1);
});

File: backend-py/seed.py
from __future__ import annotations

import sys
from datetime import date, datetime, timedelta, timezone
from decimal import Decimal
from random import Random
from uuid import uuid4

from faker import Faker

from app.config import config
from app.constants import COUNTRY_CODES, COUNTRY_CURRENCY, DEPARTMENTS
from app.db import create_pool, migrate
from app.repos.pg import PgFxRepository
from app.services.fx import FxService

TOTAL = 10_000
SEED = 42

COUNTRY_BASE = {
    "US": "120000",
    "GB": "75000",
    "DE": "80000",
    "NL": "78000",
    "SG": "110000",
    "AU": "115000",
    "CA": "105000",
    "JP": "8500000",
    "AE": "220000",
    "IN": "1800000",
}

DEPT_FACTOR = {
    "Engineering": "1.15",
    "Product": "1.10",
    "Design": "1.00",
    "Sales": "1.05",
    "Marketing": "0.95",
    "Finance": "1.00",
    "HR": "0.90",
    "Operations": "0.85",
    "Customer Success": "0.88",
    "Legal": "1.12",
}

TITLES = {
    "Engineering": ["Software Engineer", "Senior Engineer", "Staff Engineer", "Engineering Manager"],
    "Product": ["Product Manager", "Senior Product Manager", "Product Lead"],
    "Design": ["Product Designer", "Senior Designer", "Design Lead"],
    "Sales": ["Account Executive", "Senior AE", "Sales Manager"],
    "Marketing": ["Marketing Manager", "Content Lead", "Growth Manager"],
    "Finance": ["Financial Analyst", "Accountant", "Finance Manager"],
    "HR": ["HR Generalist", "HRBP", "People Partner"],
    "Operations": ["Operations Associate", "Ops Manager"],
    "Customer Success": ["CSM", "Senior CSM"],
    "Legal": ["Counsel", "Senior Counsel"],
}

COUNTRY_WEIGHTS = [28, 22, 10, 8, 6, 6, 6, 5, 5, 4]


def add_days(iso: str, days: int) -> str:
    parsed = date.fromisoformat(iso) + timedelta(days=days)
    return parsed.isoformat()


def amount(country: str, department: str, rng: Random) -> Decimal:
    mid = Decimal(COUNTRY_BASE[country]) * Decimal(DEPT_FACTOR[department])
    factor = Decimal(str(round(0.72 + rng.random() * 0.73, 4)))
    return (mid * factor).quantize(Decimal("1"))


def insert_batch(conn, sql: str, rows: list[tuple], batch: int = 200) -> None:
    for i in range(0, len(rows), batch):
        slice_rows = rows[i : i + batch]
        placeholders = ",".join(["(" + ",".join(["%s"] * len(slice_rows[0])) + ")"] * len(slice_rows))
        params: list = []
        for row in slice_rows:
            params.extend(row)
        conn.execute(f"{sql} VALUES {placeholders}", params)
    conn.commit()


def refuse_remote_seed(url: str) -> None:
    lowered = url.lower()
    if "neon.tech" in lowered or "neon.database" in lowered:
        print("Refusing to seed a Neon database. Use local Docker Postgres.")
        sys.exit(1)


def run() -> None:
    refuse_remote_seed(config.database_url)
    pool = create_pool()
    migrate(pool)
    force = "--force" in sys.argv
    with pool.connection() as conn:
        count = conn.execute("SELECT COUNT(*)::int AS count FROM employees").fetchone()[0]
        if count > 0 and not force:
            print("Database already has employees. Refusing to seed twice. Pass --force to replace.")
            pool.close()
            return
        if force:
            conn.execute("TRUNCATE salary_records, employees CASCADE")
            conn.commit()

    FxService(PgFxRepository(pool)).seed_defaults()

    rng = Random(SEED)
    faker = Faker()
    faker.seed_instance(SEED)
    as_of = "2026-08-29"
    weight_sum = sum(COUNTRY_WEIGHTS)
    employees: list[tuple] = []
    salaries: list[tuple] = []
    now = datetime.now(timezone.utc)

    for index in range(1, TOTAL + 1):
        country_roll = rng.random() * weight_sum
        acc = 0.0
        country = COUNTRY_CODES[0]
        for i, code in enumerate(COUNTRY_CODES):
            acc += COUNTRY_WEIGHTS[i]
            if country_roll <= acc:
                country = code
                break
        department = DEPARTMENTS[rng.randrange(len(DEPARTMENTS))]
        employee_id = str(uuid4())
        employees.append(
            (
                employee_id,
                f"ACME-{index:05d}",
                faker.name(),
                f"{faker.user_name()}.{index}@acme.test".lower(),
                country,
                department,
                TITLES[department][rng.randrange(len(TITLES[department]))],
                "inactive" if rng.random() < 0.08 else "active",
                now,
                now,
            )
        )
        if rng.random() < 0.02:
            continue
        hired = add_days(as_of, -int(30 + rng.random() * 365 * 8))
        currency = COUNTRY_CURRENCY[country]
        last_from = add_days(hired, int(rng.random() * 60))
        last_amount = amount(country, department, rng)
        salaries.append((str(uuid4()), employee_id, f"{last_amount:.2f}", currency, last_from, "offer accepted", now))
        extra = 0 if rng.random() < 0.55 else (1 if rng.random() < 0.78 else 2)
        for _ in range(extra):
            last_from = add_days(last_from, int(280 + rng.random() * 140))
            if last_from > add_days(as_of, 180):
                break
            last_amount = (last_amount * Decimal(str(round(1.04 + rng.random() * 0.14, 4)))).quantize(Decimal("1"))
            salaries.append(
                (
                    str(uuid4()),
                    employee_id,
                    f"{last_amount:.2f}",
                    currency,
                    last_from,
                    "scheduled raise" if last_from > as_of else ["annual revision", "promotion", "market adjustment"][rng.randrange(3)],
                    now,
                )
            )

    with pool.connection() as conn:
        insert_batch(
            conn,
            "INSERT INTO employees (id, employee_code, name, email, country_code, department, designation, status, created_at, updated_at)",
            employees,
        )
        insert_batch(
            conn,
            "INSERT INTO salary_records (id, employee_id, base_amount, currency, effective_from, reason, created_at)",
            salaries,
        )
    print(f"Seeded {len(employees)} employees and {len(salaries)} salary records.")
    pool.close()


if __name__ == "__main__":
    run()
